turbopack/crates/turbo-tasks-fs/src/watcher.rs

```
@@ -152,60 +176,72 @@ impl NonRecursiveDiskWatcherState {
+        let Some(internal) = &mut *internal_guard else {
+            return Ok(());
+        };
         if self.watching.insert(dir_path.to_path_buf()) {
-            self.start_watching_dir(&mut internal, dir_path, root_path)?;
+            self.start_watching_dir_and_parents(internal, dir_path, root_path)?;
         }
         Ok(())
     }
 
-    /// Private helper, assumes that the path has already been added to `self.watching`.
+    /// Private helper, assumes that `dir_path` has already been added to `self.watching`.
+    ///
+    /// This does not watch any of the parent directories. For that, use
+    /// [`start_watching_dir_and_parents`]. Use this method when iterating over previously-watched
+    /// values in `self.watching`.
     fn start_watching_dir(
         &self,
-        watcher_internal_guard: &mut MutexGuard<Option<DiskWatcherInternal>>,
+        watcher_internal: &mut DiskWatcherInternal,
         dir_path: &Path,
         root_path: &Path,
     ) -> Result<()> {
         debug_assert_ne!(dir_path, root_path);
-        let Some(watcher_internal_guard) = watcher_internal_guard.as_mut() else {
-            return Ok(());
-        };
 
-        let mut path = dir_path;
-        let err_with_context = |err: anyhow::Error| {
-            return Err(err).context(format!(
-                "Unable to watch {} (tried up to {})",
-                dir_path.display(),
-                path.display()
-            ));
-        };
+        match watcher_internal.watch(dir_path, RecursiveMode::NonRecursive) {
+            Ok(())
+            | Err(notify::Error {
+                // The path was probably deleted before we could process the event, but the parent
+                // should still be watched. The codepaths that care about this either call
+                // `start_watching_dir_and_parents` or handle the parents themselves.
+                kind: notify::ErrorKind::PathNotFound,
```

I wasn't entirely sure, but will this hide errors if there isn't a parent directory any more? What happens in this case? Is that handled by the wording: "or handle the parents themselves"?

---

This can happen when directories are rapidly added and removed, and we should not generate any sort of error in that case. It can also occur when trying to read a file from a directory that doesn't exist, but the error there should come from the read operation, not the watch operation. In both these cases, we just want to know if the file or directory gets created again later so that we can invalidate the read (e.g. so that it no longer returns an error).

We assume that the `root_path` is never deleted, so as long as code traverses and tries to add all parents until it finds the `root_path`, we're guaranteed to at least watch *something*.

The codepaths that call `start_watching_dir` are `start_watching_dir_and_parents`, `restore_if_watching`, and `restore_all_watching`.

- `start_watching_dir_and_parents` explicitly adds the parents.
- `restore_if_watching` gets called as the result of a notify event from the parent directory, so we can assume we're already watching the parent when it's called.
- `restore_all_watching` adds every value in `watching`. One of the invariants of `watching` (documented on the type declaration) is that every path in `watching` also has it's parents in `watching`.

I agree, this code is really hard to follow, but I also don't want to have paragraphs of comments on every line of code...

---

I think there is a race condition here, when the child dir is created while we are walking the parent directories.

1. we try to watch /path/to/dir but it doesn't exist.
2. /path/to/dir is created
3. we move to the parent directory (`path/to`) and start watching it.
4. But we already missed the event that path/to/dir was created.

I think we need to run the loop reverse direction, watching parent directories first before watching the child directories. This way we already have the parent directory watcher setup before "reading" the child directory (reading in the sense of starting to watch it).

---

Fixed in https://github.com/vercel/next.js/pull/82454